File: Days 31 - 40/Day 34/GetMinimumPalindrome.cpp
#include <algorithm>
#include <iostream>
#include <string>

inline bool IsPalindrome(const std::string& string)
{
	std::string reversedString = string;
	std::reverse(reversedString.begin(), reversedString.end());

	return string == reversedString;
}

std::string GetMinimumPalindrome(const std::string& string)
{
	if (IsPalindrome(string))
	{
		return string;
	}

	if (string[0] == string[string.length() - 1])
	{
		return string[0] + GetMinimumPalindrome(string.substr(1, string.length() - 2)) + string[string.length() - 1];
	}
	else
	{
		std::string firstPalindrome = string[0] + GetMinimumPalindrome(string.substr(1)) + string[0];
		std::string secondPalindrome = string[string.length() - 1] + GetMinimumPalindrome(string.substr(0, string.length() - 1)) + string[string.length() - 1];

		if (firstPalindrome.length() > secondPalindrome.length())
		{
			return secondPalindrome;
		}
		else if (firstPalindrome.length() < secondPalindrome.length())
		{
			return firstPalindrome;
		}
		else
		{
			return (firstPalindrome < secondPalindrome) ? firstPalindrome : secondPalindrome;
		}
	}
}

int main(int argc, const char* argv[])
{
	std::cout << GetMinimumPalindrome("racecar") << "\n";
	std::cout << GetMinimumPalindrome("race") << "\n";
	std::cout << GetMinimumPalindrome("google") << "\n";

	std::cin.get();

	return 0;
}

File: Days 31 - 40/Day 34/GetMinimumPalindrome.cs
using System;

namespace DailyCodingProblem
{
	internal class Day34
	{
		private static int Main(string[] args)
		{
			Console.WriteLine(GetMinimumPalindrone("racecar"));
			Console.WriteLine(GetMinimumPalindrone("race"));
			Console.WriteLine(GetMinimumPalindrone("google"));

			Console.ReadLine();

			return 0;
		}

		private static string GetMinimumPalindrone(string text)
		{
			if (IsPalindrome(text))
			{
				return text;
			}

			if (text[0] == text[text.Length - 1])
			{
				return text[0] + GetMinimumPalindrone(text.Substring(1, text.Length - 2)) + text[text.Length - 1];
			}

			string firstPalindrome = text[0] + GetMinimumPalindrone(text.Substring(1)) + text[0];
			string secondPalindrome = text[text.Length - 1] + GetMinimumPalindrone(text.Substring(0, text.Length - 1)) + text[text.Length - 1];

			if (firstPalindrome.Length > secondPalindrome.Length)
			{
				return secondPalindrome;
			}
			else if (firstPalindrome.Length < secondPalindrome.Length)
			{
				return firstPalindrome;
			}
			else
			{
				return string.Compare(firstPalindrome, secondPalindrome) <= 0 ? firstPalindrome : secondPalindrome;
			}
		}

		private static bool IsPalindrome(string text)
		{
			char[] chars = text.ToCharArray();
			Array.Reverse(chars);

			return text == new string(chars);
		}
	}
}